x-pack/plugins/elastic_assistant/server/lib/model_evaluator/evaluation.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { loadEvaluator } from 'langchain/evaluation';
+import { LLM } from 'langchain/llms/base';
+import { ChainValues, HumanMessage } from 'langchain/schema';
+
+import { chunk } from 'lodash/fp';
+
+import { Logger } from '@kbn/core/server';
+
+import { ToolingLog } from '@kbn/tooling-log';
+import { asyncForEach } from '@kbn/std';
+import { AgentExecutorEvaluator } from '../langchain/executors/types';
+import { Dataset } from '../../schemas/evaluate/post_evaluate';
+import { getMessageFromLangChainResponse, wait } from './utils';
+import { ResponseBody } from '../langchain/helpers';
+
+export interface PerformEvaluationParams {
+  agentExecutorEvaluators: AgentExecutorEvaluator[];
+  dataset: Dataset;
+  evaluatorModel: LLM;
+  evaluationPrompt?: string;
+  evaluationType: string;
+  maxConcurrency?: number;
+  logger: Logger | ToolingLog;
+}
+
+export interface EvaluationResult {
+  evaluation: ChainValues;
+  input: string;
+  prediction: string;
+  predictionResponse: ResponseBody;
+  reference: string;
+}
+
+/**
+ * Evaluates a dataset based on an evaluation rubric. Takes dataset of input/reference pairs,
+ * and fetches the output (prediction) of the input against the provided agent executors.
+ * Then evaluates all three using the provided evaluation rubric.
+ */
+export const performEvaluation = async ({
+  agentExecutorEvaluators,
+  dataset,
+  evaluatorModel,
+  evaluationPrompt,
+  evaluationType,
+  maxConcurrency = 3,
+  logger,
+}: PerformEvaluationParams) => {
+  const results: ChainValues = [];
+  const requests = dataset.flatMap(({ input }) =>
+    agentExecutorEvaluators.map((agent) => agent([new HumanMessage(input)]))
+  );
+
+  logger.info(`Total Requests: ${requests.length}`);
+
+  logger.info('Fetching predictions...');
+  const requestChunks = chunk(maxConcurrency, requests);
+  for (let i = 0; i < requestChunks.length; i++) {
+    logger.info(`Request chunk [${i}]: ${requestChunks[i].length}`);
+
+    const chunkResults = await Promise.all(requestChunks[i]);
+    logger.info(`Chunk Results:\n${JSON.stringify(chunkResults)}`);
+    results.push(...chunkResults);
+  }
+
+  const mergedResults = dataset.map((d, i) => ({
+    ...d,
+    prediction: getMessageFromLangChainResponse(results[i]),
+    predictionResponse: results[i],
```

Comment:
does this visit every entry?

Reply:
It does not -- thank you! Will get a fix in for this 🙂